### PR TITLE
[WIP] Fix negative intervals

### PR DESCRIPTION
**Pull request by trangmei**
Implement interim solutions to fix https://github.com/Coursemology/coursemology2/issues/3234.

# TODO:
- [ ] Reviewer to comment if this should be pursued @jeremyyap 
- [ ] If yes, write tests to simulate edge cases and cover new functions

### CODE REVIEW

**Comment by trangmei** [comment]
The cause behind negative intervals is seemingly 'corrupt' video events data, possibly caused by either frontend performance (causing loss in video events capturing) or a glitch in upsert.
I'll need some time to do a deep dive into these components to first reproduce the 'corrupt' data (still can't reproduce). There's a real chance that even when problem can be reproduced and root cause found, there's no fix.

The hacky 'solutions' implemented in this PR only treat the symptoms of the problem, by tweaking the way we interpret raw video events data. No changes is made to the way this data is recorded. So regardless of whether the problem can be reproduced, root cause found and fixed, I think it's still good to have these fixes in place.

File: app/models/concerns/course/video/watch_statistics_concern.rb [hunk 1]
```diff
@@ -33,6 +33,8 @@ def watch_frequency
 
   private
 
+  EVENT_TYPES = { start_times: ['play', 'seek_end'], end_times: ['pause', 'seek_start', 'end'] }.freeze
```
[the remark this answers is not in the document]
Comment (trangmei) [reply app/models/concerns/course/video/watch_statistics_concern.rb]: The keys are shared between (1) hashes containing arrays of Course::Video::Events and (2) hashes containing arrays of events' video_time (integer). So if I change hash keys i have to change all of them, and then it wouldn't make sense for the second type of hashes. (See `filter_interval_events` definition.)

We can consider changing it to `:starts` and `:ends` so that it sounds relevant to both types of hashes.

[the remark this answers is not in the document]
Comment (fonglh) [reply app/models/concerns/course/video/watch_statistics_concern.rb]: `start` and `end` then.

File: spec/factories/course_video_sessions.rb [hunk 2]
```diff
@@ -83,5 +83,42 @@
...
+                                              video_time: 0)
+        session.events << build(:video_event, sequence_num: 2, event_type: 'end',
+                                              video_time: 70)
+        session.events << build(:video_event, sequence_num: 3, event_type: 'play',
+                                              video_time: 70)
+        session.events << build(:video_event, sequence_num: 4, event_type: 'buffer',
+                                              video_time: 70)
+        session.last_video_time = 4
+      end
+    end
```
Comment (jeremyyap) [inline spec/factories/course_video_sessions.rb]: Maybe we could clean up this file a little?

Something like

```
event_params = [
  { event_type: 'play', video_time: 0 },
  { event_type: 'end', video_time: 70),
  ....
]
session.events = event_param.map { |hash| build(:video_event, params) }
```

This would make it easier to see the set of events generated by using each trait.

As for the sequence number, since it is usually sequential we could use a sequence in the `course_video_events` factory.

```
  sequence :sequence_num
```

Other than that LGTM

Comment (trangmei) [reply spec/factories/course_video_sessions.rb]: I can't find a way to rewind the sequence for every new session. Factory_bot only has rewind_sequences, which rewinds all sequences.
So i'll clean this file up as suggested, but still supply :sequence_num as a param to build.
